### PR TITLE
feat: update es with dataservices objects

### PR DESCRIPTION
**Pull request by geoffreyaldebert**
Issue https://github.com/datagouv/data.gouv.fr/issues/1405

### CODE REVIEW

File: udata_search_service/infrastructure/search_clients.py [hunk 1]
```diff
@@ -339,6 +359,54 @@ def query_reuses(self, query_text: str, offset: int, page_size: int, filters: di
         res = [hit.to_dict(skip_empty=False) for hit in response.hits]
         return results_number, res
 
+    def query_dataservices(self, query_text: str, offset: int, page_size: int, filters: dict, sort: Optional[str] = None) -> Tuple[int, List[dict]]:
+        s = SearchableDataservice.search()
+
+        for key, value in filters.items():
+            s = s.filter('term', **{key: value})
+
+        dataservices_score_functions = [
```
Comment (maudetes) [inline udata_search_service/infrastructure/search_clients.py:368]: We can update the score functions factors depending on search results in production

**Review by maudetes — APPROVED** [review]
I think we have a good first iteration for dataservice search! :rocket:

File: udata_search_service/infrastructure/search_clients.py [hunk 2]
```diff
@@ -339,6 +359,54 @@ def query_reuses(self, query_text: str, offset: int, page_size: int, filters: di
         res = [hit.to_dict(skip_empty=False) for hit in response.hits]
         return results_number, res
 
+    def query_dataservices(self, query_text: str, offset: int, page_size: int, filters: dict, sort: Optional[str] = None) -> Tuple[int, List[dict]]:
+        s = SearchableDataservice.search()
```
Comment (magopian) [inline udata_search_service/infrastructure/search_clients.py:363]: Nit: could we possible not use single letter variables? Maybe something like `search`?

File: udata_search_service/presentation/api.py [hunk 3]
```diff
@@ -93,6 +93,27 @@ def sort_validate(cls, value):
...
+
+    @validator('sort')
+    def sort_validate(cls, value):
+        sorts = [
+            'created',
+        ]
+        choices = sorts + ['-' + k for k in sorts]
+        if value not in choices:
+            raise ValueError('Sort parameter is not in the sorts available choices.')
+        return value
```
Comment (magopian) [inline udata_search_service/presentation/api.py:114]: Nit: is this for future sorts? It feels a bit overkill for the moment 🤣

File: udata_search_service/presentation/api.py [hunk 4]
```diff
@@ -346,6 +367,78 @@ def reuse_unindex(reuse_id: str, reuse_service: ReuseService = Provide[Container
...
+    try:
+        request_args = DataserviceArgs(**request.args)
+    except ValidationError as e:
+        abort(400, e)
+
+    results, results_number, total_pages = dataservice_service.search(request_args.dict())
+
+    next_url = url_for('api.dataservice_search', q=request_args.q, page=request_args.page + 1,
+                       page_size=request_args.page_size, _external=True)
+    prev_url = url_for('api.dataservice_search', q=request_args.q, page=request_args.page - 1,
```
Comment (magopian) [inline udata_search_service/presentation/api.py:382]: Nit: should this be clamped to make sure we never get a `page=0` or worse, a negative one?

File: udata_search_service/presentation/api.py [hunk 5]
```diff
@@ -346,6 +367,78 @@ def reuse_unindex(reuse_id: str, reuse_service: ReuseService = Provide[Container
...
+@inject
+def dataservices_search(dataservice_service: DataserviceService = Provide[Container.dataservice_service]):
+    try:
+        request_args = DataserviceArgs(**request.args)
+    except ValidationError as e:
+        abort(400, e)
+
+    results, results_number, total_pages = dataservice_service.search(request_args.dict())
+
+    next_url = url_for('api.dataservice_search', q=request_args.q, page=request_args.page + 1,
```
Comment (magopian) [inline udata_search_service/presentation/api.py:380]: Nit: should this be checked so we return `None` when `request_args.page == total_pages`?

**Review by magopian — APPROVED** [review]
Wow, that was a big one, congratz!
Good job, a couple of nits that you man want to ignore ;)

File: udata_search_service/presentation/api.py (hunk 4 above)
Comment (maudetes) [reply udata_search_service/presentation/api.py:382]: I think edge cases are dealth in [make_response](https://github.com/opendatateam/udata-search-service/blob/04b0be7d247eebb1b44e322fb39af183cf9bec55/udata_search_service/presentation/api.py#L96)

File: udata_search_service/presentation/api.py (hunk 5 above)
Comment (maudetes) [reply udata_search_service/presentation/api.py:380]: I think edge cases are dealth in [make_response](https://github.com/opendatateam/udata-search-service/blob/04b0be7d247eebb1b44e322fb39af183cf9bec55/udata_search_service/presentation/api.py#L96)

File: udata_search_service/domain/factories.py [hunk 6]
```diff
@@ -73,3 +73,19 @@ class Meta:
...
+    id = factory.Faker('md5')
+    title = factory.Faker('sentence')
+    description = factory.Faker('text')
+    description_length = factory.Faker("random_int")
+    created_at = factory.LazyFunction(datetime.datetime.utcnow)
+    followers = factory.Faker("random_int")
+    is_restricted = factory.Faker("boolean")
+    organization = factory.Faker('md5')
+    organization_name = factory.Faker('company')
+    owner = factory.Faker('md5')
```
[the remark this answers is not in the document]
Comment (maudetes) [reply udata_search_service/domain/factories.py:91]: Indeed, it was used all around already

File: udata_search_service/presentation/api.py (hunk 3 above)
Comment (magopian) [reply udata_search_service/presentation/api.py:114]: [sorts](https://github.com/opendatateam/udata-search-service/pull/48/files/82debfdfe1ab2711a586805c75ae901829cd3816#diff-c82642483ae4c76bf0685cb300bbea5848e8a4f8a8cc710358d52843756d2406R108-R110) are just one element, `created`, right? So this whole code could be shortened to

```python
sorts = ["created", "-created"]
if value not in sorts:
...
```

No big deal, just looked a bit funny ;)

File: tests/test_consumers.py [hunk 7]
```diff
@@ -112,6 +112,60 @@ def test_parse_reuse_obj():
...
+    # Make sure that these fields are loaded as is
+    for key in ['id', 'title', 'tags', 'owner', 'is_restricted']:
+        assert document[key] == obj[key]
+
+    # Make sure that markdown fields are stripped
+    assert document["description"] == mdstrip(obj["description"])
+
+    # Make sure that these fields are log2p-normalized
+    for key in ['followers']:
+        assert document[key] == log2p(obj[key])
```
[the remark this answers is not in the document]
Comment (magopian) [reply tests/test_consumers.py:159]: ah, I see now, thanks for the explanation, makes sense ;)

File: tests/test_api.py [hunk 8]
```diff
@@ -348,6 +348,119 @@ def test_api_reuse_index_on_another_index(app, client, search_client, faker):
...
+
+    time.sleep(2)
+
+    dataservice_resp = client.get(url_for('api.dataservice_get_specific', dataservice_id=dataservice['id']))
+    assert dataservice_resp.status_code == 200
+    assert dataservice_resp.json['title'] == dataservice['title']
+
+    dataservice_search_resp = client.get(url_for('api.dataservice_search'))
+    assert len(dataservice_search_resp.json['data']) == 1
+    assert dataservice_search_resp.json['next_page'] is None
```
[the remark this answers is not in the document]
Comment (maudetes) [reply tests/test_api.py:399]: Done in https://github.com/opendatateam/udata-search-service/pull/48/commits/03ab0f9446a98d38b7c0407b73e97ddc5c2599e9

File: tests/test_api.py [hunk 9]
```diff
@@ -348,6 +348,119 @@ def test_api_reuse_index_on_another_index(app, client, search_client, faker):
     assert resp['_source']['title'] == reuse['title']
 
 
+def test_api_dataservice_index_unindex(app, client, search_client, faker):
```
[the remark this answers is not in the document]
Comment (maudetes) [reply tests/test_api.py]: Removed in https://github.com/opendatateam/udata-search-service/pull/48/commits/b3fc7bb7ba06afb69b5567ea175c6556a3378be9

File: tests/test_search_client.py [hunk 10]
```diff
@@ -307,6 +359,8 @@ def test_general_search_with_sorting(app, client, search_client, faker):
     assert res[0]['name'] == 'org-test-2'
     results_number, res = search_client.query_reuses(None, 0, 20, {}, sort='-followers')
     assert res[0]['title'] == 'reuse-test-2'
+    results_number, res = search_client.query_dataservices(None, 0, 20, {}, sort='-followers')
+    assert res[0]['title'] == 'dataservice-test-2'
```
[the remark this answers is not in the document]
Comment (maudetes) [reply tests/test_search_client.py:374]: Done in https://github.com/opendatateam/udata-search-service/pull/48/commits/130f86262604c47be428094b9c61f21ab16aba9f

**Comment by maudetes** [comment]
I think we don't want to refactor the code too much since we're targeting a major refacto of this.
Thus I would not modify styling or tests for now if it's okay with you.
See https://github.com/opendatateam/udata/pull/3109

**Comment by magopian** [comment]
Completely agree! 🚢

File: udata_search_service/presentation/api.py (hunk 3 above)
Comment (maudetes) [reply udata_search_service/presentation/api.py:114]: Indeed, it's the same logic in other models, Datasets having more sorts for example

**Pull request merged**